File: objc/test/objc/Ice/invoke/BlobjectI.m
// **********************************************************************
//
// Copyright (c) 2003-2011 ZeroC, Inc. All rights reserved.
//
// This copy of Ice Touch is licensed to you under the terms described in the
// ICE_TOUCH_LICENSE file included in this distribution.
//
// **********************************************************************

#import <Ice/Ice.h>
#import <invoke/BlobjectI.h>

@implementation BlobjectI
-(BOOL) ice_invoke:(NSData*)inParams outParams:(NSMutableData**)outParams current:(ICECurrent*)current
{
    id<ICECommunicator> communicator = [current.adapter getCommunicator];
    id<ICEInputStream> inS = [ICEUtil createInputStream:communicator data:inParams];
    id<ICEOutputStream> outS = [ICEUtil createOutputStream:communicator];
    if([current.operation isEqualToString:@"opOneway"])
    {
        return YES;
    }
    else if([current.operation isEqualToString:@"opString"])
    {
        NSString* s = [inS readString];
        [outS writeString:s];
        [outS writeString:s];
        *outParams = [outS finished];
        return YES;
    }
    else if([current.operation isEqualToString:@"opException"])
    {
        TestInvokeMyException* ex = [TestInvokeMyException myException];
        [outS writeException:ex];
        *outParams = [outS finished];
        return NO;
    }
    else if([current.operation isEqualToString:@"shutdown"])
    {
        [communicator shutdown];
        return YES;
    }
    else if([current.operation isEqualToString:@"ice_isA"])
    {
        NSString* s = [inS readString];
        if([s isEqualToString:[TestInvokeMyClass ice_staticId]])
        {
            [outS writeBool:YES];
        }
        else
        {
            [outS writeBool:NO];
        }
        *outParams = [outS finished];
        return YES;
    }
    else
    {
        @throw [ICEOperationNotExistException operationNotExistException:__FILE__ 
                                              line:__LINE__
                                              id_:current.id_
                                              facet:current.facet
                                              operation:current.operation];
    }
}
@end

File: objc/test/cpp/Ice/invoke/BlobjectI.cpp
// **********************************************************************
//
// Copyright (c) 2003-2011 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_TOUCH_LICENSE file included in this distribution.
//
// **********************************************************************

#include <Ice/Ice.h>
#include <invoke/BlobjectI.h>
#include <InvokeTest.h>
#include <TestCommon.h>

using namespace std;
using namespace Test::Invoke;

bool
invokeInternal(const Ice::InputStreamPtr& in, vector<Ice::Byte>& outParams, const Ice::Current& current)
{
    Ice::CommunicatorPtr communicator = current.adapter->getCommunicator();
    Ice::OutputStreamPtr out = Ice::createOutputStream(communicator);
    if(current.operation == "opOneway")
    {
        return true;
    }
    else if(current.operation == "opString")
    {
        string s;
        in->read(s);
        out->write(s);
        out->write(s);
        out->finished(outParams);
        return true;
    }
    else if(current.operation == "opException")
    {
        Test::Invoke::MyException ex;
        out->writeException(ex);
        out->finished(outParams);
        return false;
    }
    else if(current.operation == "shutdown")
    {
        communicator->shutdown();
        return true;
    }
    else if(current.operation == "ice_isA")
    {
        string s;
        in->read(s);
        if(s == "::Test::Invoke::MyClass")
        {
            out->write(true);
        }
        else
        {
            out->write(false);
        }
        out->finished(outParams);
        return true;
    }
    else
    {
        Ice::OperationNotExistException ex(__FILE__, __LINE__);
        ex.id = current.id;
        ex.facet = current.facet;
        ex.operation = current.operation;
        throw ex;
    }
}

bool
BlobjectI::ice_invoke(const vector<Ice::Byte>& inParams, vector<Ice::Byte>& outParams, const Ice::Current& current)
{
    Ice::InputStreamPtr in = Ice::createInputStream(current.adapter->getCommunicator(), inParams);
    return invokeInternal(in, outParams, current);
}
